### PR TITLE
MPS files when a daily optimization is asked

### PR DESCRIPTION
**Pull request by guilpier-code**
Before this fix, when user asked for a daily optimization, only 2 MPS files used to be created per week (**problem-\<year\>-\<week\>--optim-nb-\<opt-number\>.mps**) .
This is because MPS files generated by some daily optimization were overridden by files generated by the next daily optimization : indeed, they had the same name.
Now, MPS files wear the name : **problem-\<year\>-\<week\>-\<day\>--optim-nb-\<opt-number\>.mps** so they cannot be overridden.
Note that what we say here about MPS files is also true for criterion files (ex : **criterion-1-2-6--optim-nb-1.txt**)

Besides, some preliminary refactoring works were made to ease the current work. As usual, associated branches are sitting on top of each other : 
- #1030 
- #1032
- #1040

### CODE REVIEW

File: src/solver/utils/filename.h [hunk 1]
```diff
@@ -1,10 +1,57 @@
...
+public:
+    optWeekly(unsigned int week, unsigned int year);
+    std::string to_string() const override;
+
+private:
+    unsigned int week_ = 0;
+    unsigned int year_ = 0;
+};
+
+std::shared_ptr<optPeriod> createOptimizationPeriod(bool isOptimizationWeekly,
```
Comment (flomnes) [inline src/solver/utils/filename.h:47]:
```suggestion
std::shared_ptr<optPeriod> createOptimizationPeriod(Antares::Data::SimplexOptimization optPeriod,
```

File: src/solver/utils/filename.h [hunk 2]
```diff
@@ -1,10 +1,57 @@
...
-std::string getFilenameWithExtension(const YString& prefix,
-                                     const YString& extension,
-                                     uint year,
-                                     uint week,
-                                     uint optNumber = 0);
+// ------------------------------------
+// Optimization period : base class
+// ------------------------------------
+class optPeriod
+{
```
Comment (flomnes) [inline src/solver/utils/filename.h]: // This class intends to provide an interface for the naming of optimization-related files (MPS files, etc.)

File: src/solver/utils/mps_utils.cpp [hunk 3]
```diff
@@ -324,58 +321,41 @@ void OPT_EcrireJeuDeDonneesLineaireAuFormatMPS(
...
-                             uint year,
-                             uint week,
                              uint optNumber) :
-    I_MPS_writer(year, week, optNumber),
+    I_MPS_writer(optNumber),
     named_splx_problem_(named_splx_problem)
 {}
 
-void fullMPSwriter::runIfNeeded(Solver::IResultWriter::Ptr writer)
+void fullMPSwriter::runIfNeeded(Solver::IResultWriter::Ptr writer, std::string filename)
```
Comment (flomnes) [inline src/solver/utils/mps_utils.cpp]:
```suggestion
void fullMPSwriter::runIfNeeded(Solver::IResultWriter::Ptr writer, const std::string& filename)
```

File: src/solver/utils/mps_utils.cpp [hunk 4]
```diff
@@ -324,58 +321,41 @@ void OPT_EcrireJeuDeDonneesLineaireAuFormatMPS(
...
-                                           uint year,
-                                           uint week,
                                            uint optNumber) :
-    I_MPS_writer(year, week, optNumber),
+    I_MPS_writer(optNumber),
     solver_(solver)
 {
 }
-void fullOrToolsMPSwriter::runIfNeeded(Solver::IResultWriter::Ptr writer)
+void fullOrToolsMPSwriter::runIfNeeded(Solver::IResultWriter::Ptr writer, std::string filename)
```
Comment (flomnes) [inline src/solver/utils/mps_utils.cpp]:
```suggestion
void fullOrToolsMPSwriter::runIfNeeded(Solver::IResultWriter::Ptr writer, const std::string& filename)
```

File: src/solver/utils/mps_utils.h [hunk 5]
```diff
@@ -17,10 +17,8 @@ using namespace Antares::Optimization;
 using namespace operations_research;
 
 void OPT_EcrireJeuDeDonneesLineaireAuFormatMPS(void*,
-                                               uint,
-                                               uint,
-                                               uint,
-                                               Solver::IResultWriter::Ptr writer);
+                                               Solver::IResultWriter::Ptr writer,
+                                               std::string filename);
```
Comment (flomnes) [inline src/solver/utils/mps_utils.h]:
```suggestion
                                               const std::string& filename);
```

File: src/solver/optimisation/opt_appel_solveur_lineaire.cpp [hunk 6]
```diff
@@ -243,11 +244,14 @@ bool OPT_AppelDuSimplexe(PROBLEME_HEBDO* ProblemeHebdo,
     {
         solver = ORTOOLS_ConvertIfNeeded(&Probleme, solver);
     }
-
-    mpsWriterFactory mps_writer_factory(
-      ProblemeHebdo, NumIntervalle, numeroOptimisation, &Probleme, ortoolsUsed, solver);
+    std::string filename = createMPSfilename(opt_period, numeroOptimisation);
```
Comment (flomnes) [inline src/solver/optimisation/opt_appel_solveur_lineaire.cpp]:
```suggestion
    const std::string filename = createMPSfilename(opt_period, numeroOptimisation);
```

File: src/solver/optimisation/opt_fonctions.h [hunk 7]
```diff
@@ -60,7 +61,7 @@ bool OPT_PilotageOptimisationQuadratique(PROBLEME_HEBDO*);
 **
 ** \return True si l'operation s'est bien deroulee, false si le probleme n'a pas de solution
 */
-bool OPT_AppelDuSimplexe(PROBLEME_HEBDO*, int, const int);
+bool OPT_AppelDuSimplexe(PROBLEME_HEBDO*, int, const int, std::shared_ptr<optPeriod>);
```
Comment (flomnes) [inline src/solver/optimisation/opt_fonctions.h]: Add explanations on this new argument (daily/weekly opt.)

**Review by flomnes — CHANGES_REQUESTED** [review]

File: src/solver/utils/mps_utils.cpp (hunk 3 above)
Comment (guilpier-code) [reply src/solver/utils/mps_utils.cpp]: Done

File: src/solver/utils/mps_utils.cpp (hunk 4 above)
Comment (guilpier-code) [reply src/solver/utils/mps_utils.cpp]: Done

File: src/solver/utils/mps_utils.h (hunk 5 above)
Comment (guilpier-code) [reply src/solver/utils/mps_utils.h]: Done

File: src/solver/optimisation/opt_appel_solveur_lineaire.cpp (hunk 6 above)
Comment (guilpier-code) [reply src/solver/optimisation/opt_appel_solveur_lineaire.cpp]: Done

File: src/solver/utils/filename.h (hunk 2 above)
Comment (guilpier-code) [reply src/solver/utils/filename.h]: Done

File: src/solver/optimisation/opt_fonctions.h (hunk 7 above)
Comment (guilpier-code) [reply src/solver/optimisation/opt_fonctions.h]: Done

File: src/solver/utils/filename.h [hunk 8]
```diff
@@ -1,10 +1,74 @@
...
+
+std::shared_ptr<optPeriodAsString> createOptimizationPeriod(bool isOptimizationWeekly,
+                                                    unsigned int day,
+                                                    unsigned int week,
+                                                    unsigned int year); 
+
+std::string createOptimizationFilename(const std::string& title,
+                                       std::shared_ptr<optPeriodAsString> opt_period_as_string,
+                                       const unsigned int optNumber,
+                                       const std::string& extension);
```
Comment (flomnes) [inline src/solver/utils/filename.h]: This function is only used in the implementation. IMO, it shouldn't be declared here, only in the cpp file.

File: src/solver/utils/filename.cpp [hunk 9]
```diff
@@ -1,23 +1,77 @@
 #include <sstream>
-#include <antares/study.h>
+#include <antares/logs/logs.h>
```
Comment (flomnes) [inline src/solver/utils/filename.cpp]:
```suggestion
#include <antares/logs.h>
```

File: src/solver/optimisation/opt_optimisation_lineaire.cpp [hunk 10]
```diff
@@ -89,13 +100,23 @@ bool OPT_OptimisationLineaire(PROBLEME_HEBDO* ProblemeHebdo, uint numSpace)
...
+        if (!OPT_AppelDuSimplexe(ProblemeHebdo, NumeroDeLIntervalle, optimizationNumber, opt_period_as_string))
             return false;
 
         if (ProblemeHebdo->ExportMPS != Data::mpsExportStatus::NO_EXPORT || ProblemeHebdo->Expansion == OUI_ANTARES)
-            OPT_EcrireResultatFonctionObjectiveAuFormatTXT(ProblemeHebdo,
-                                                           NumeroDeLIntervalle, 
-                                                           optimizationNumber);
+        {
+            double CoutOptimalSolution = OPT_ObjectiveFunctionResult(ProblemeHebdo, NumeroDeLIntervalle, optimizationNumber);
+            OPT_EcrireResultatFonctionObjectiveAuFormatTXT(CoutOptimalSolution, opt_period_as_string, optimizationNumber);
```
Comment (flomnes) [inline src/solver/optimisation/opt_optimisation_lineaire.cpp]: `CoutOptimalSolution` should be named `OptimalSolutionCost`

File: src/solver/optimisation/opt_optimisation_lineaire.cpp [hunk 11]
```diff
@@ -35,6 +35,17 @@
 using namespace Antares;
 using namespace Yuni;
 
+double OPT_ObjectiveFunctionResult(const PROBLEME_HEBDO* Probleme,
+                                   const int NumeroDeLIntervalle,
+                                   const int numeroOptimisation)
+{
+    if (numeroOptimisation == PREMIERE_OPTIMISATION)
```
Comment (flomnes) [inline src/solver/optimisation/opt_optimisation_lineaire.cpp]:
```suggestion
                                   const int optimizationNumber)
{
    if (optimizationNumber == PREMIERE_OPTIMISATION)
```

File: src/solver/optimisation/opt_optimisation_lineaire.cpp [hunk 12]
```diff
@@ -89,13 +100,23 @@ bool OPT_OptimisationLineaire(PROBLEME_HEBDO* ProblemeHebdo, uint numSpace)
         OPT_InitialiserLesCoutsLineaire(
           ProblemeHebdo, PremierPdtDeLIntervalle, DernierPdtDeLIntervalle, numSpace);
 
-        if (!OPT_AppelDuSimplexe(ProblemeHebdo, NumeroDeLIntervalle, optimizationNumber))
+        // An optimization period represents a sequence as <year>-<week> or <year>-<week>-<day>,
+        // depending whether the optimization is daily or weekly.
+        // These sequences are used when building the names of MPS or criterion files. 
+        std::shared_ptr<optPeriodAsString> opt_period_as_string = 
+            createOptimizationPeriod(ProblemeHebdo->OptimisationAuPasHebdomadaire,
```
Comment (flomnes) [inline src/solver/optimisation/opt_optimisation_lineaire.cpp]:
```suggestion
        auto periodGenerator = 
            createOptimizationPeriod(ProblemeHebdo->OptimisationAuPasHebdomadaire,
```

**Review by flomnes — CHANGES_REQUESTED** [review]

File: src/solver/utils/filename.h (hunk 8 above)
Comment (guilpier-code) [reply src/solver/utils/filename.h]: I totally agree ! I will remove **createOptimizationFilename** from **filename.h** and keep it defined it in **filename.cpp**
==> done

Comment (guilpier-code) [reply src/solver/utils/filename.h]: By the way, in **filename.h**, there is an pure interface : **optPeriodAsString**.
This  leads to include **filename.h** where this abstraction is used, that is as argument of : 
- **OPT_AppelDuSimplexe(...)**
- **OPT_EcrireResultatFonctionObjectiveAuFormatTXT(...)**

This 2 functions are declared in **opt_fonctions.h**.
The problem is that today, changing **filename.h** implies changing **opt_fonctions.h**, that is recompiling a lot of code (as **opt_fonctions.h** is included in a lot of files).
I recommend isolating **optPeriodAsString** in a **opt-period-as-string.h** to be included in **opt_fonctions.h** and **filename.h**.
The abstraction **optPeriodAsString** is not supposed to change, but its implementations are supposed to change.
So, if we do that, changing  **filename.h** won't affect **opt_fonctions.h** anymore, and recompilation time will be much shorter.

File: src/solver/utils/filename.cpp (hunk 9 above)
Comment (guilpier-code) [reply src/solver/utils/filename.cpp]: Done

File: src/solver/optimisation/opt_optimisation_lineaire.cpp (hunk 11 above)
Comment (guilpier-code) [reply src/solver/optimisation/opt_optimisation_lineaire.cpp]: Done

File: src/solver/optimisation/opt_optimisation_lineaire.cpp (hunk 12 above)
Comment (guilpier-code) [reply src/solver/optimisation/opt_optimisation_lineaire.cpp]: Done

File: src/solver/optimisation/opt_optimisation_lineaire.cpp [hunk 13]
```diff
@@ -89,13 +100,22 @@ bool OPT_OptimisationLineaire(PROBLEME_HEBDO* ProblemeHebdo, uint numSpace)
...
-        if (!OPT_AppelDuSimplexe(ProblemeHebdo, NumeroDeLIntervalle, optimizationNumber))
+        // An optimization period represents a sequence as <year>-<week> or <year>-<week>-<day>,
+        // depending whether the optimization is daily or weekly.
+        // These sequences are used when building the names of MPS or criterion files. 
+        auto opt_period_as_string = createOptimizationPeriod(ProblemeHebdo->OptimisationAuPasHebdomadaire,
+                                                             NumeroDeLIntervalle,
+                                                             ProblemeHebdo->weekInTheYear,
+                                                             ProblemeHebdo->year);
+
+        if (!OPT_AppelDuSimplexe(ProblemeHebdo, NumeroDeLIntervalle, optimizationNumber, opt_period_as_string))
```
Comment (flomnes) [inline src/solver/optimisation/opt_optimisation_lineaire.cpp]: optimizationRangeStringGenerator

File: src/solver/utils/filename.h [hunk 14]
```diff
@@ -1,10 +1,68 @@
...
+*/
+// ------------------------------------
+// Optimization period : base class
+// ------------------------------------
+class optPeriodAsString
+{
+public:
+    virtual std::string to_string() const = 0;
+    virtual ~optPeriodAsString() = default;
+};
```
Comment (flomnes) [inline src/solver/utils/filename.h]: Move this class declaration to another header file.

File: src/solver/utils/filename.h [hunk 15]
```diff
@@ -1,10 +1,68 @@
...
+class optWeeklyAsString final : public optPeriodAsString
+{
+public:
+    optWeeklyAsString(unsigned int week, unsigned int year);
+    std::string to_string() const override;
+
+private:
+    unsigned int week_ = 0;
+    unsigned int year_ = 0;
+};
```
Comment (flomnes) [inline src/solver/utils/filename.h]: Private

File: src/solver/utils/filename.h [hunk 16]
```diff
@@ -1,10 +1,68 @@
...
+
+std::shared_ptr<optPeriodAsString> createOptimizationPeriod(bool isOptimizationWeekly,
+                                                    unsigned int day,
+                                                    unsigned int week,
+                                                    unsigned int year);
```
Comment (flomnes) [inline src/solver/utils/filename.h]: Private

File: src/solver/utils/filename.h [hunk 17]
```diff
@@ -1,10 +1,68 @@
...
+
+std::string createCriterionFilename(std::shared_ptr<optPeriodAsString> opt_period_as_string, const unsigned int optNumber);
+std::string createMPSfilename(std::shared_ptr<optPeriodAsString> opt_period_as_string, const unsigned int optNumber);
```
Comment (flomnes) [inline src/solver/utils/filename.h]: Public

**Review by flomnes — CHANGES_REQUESTED** [review]

File: src/solver/utils/filename.h (hunk 8 above)
Comment (guilpier-code) [reply src/solver/utils/filename.h]: Done

File: src/solver/utils/filename.h (hunk 14 above)
Comment (guilpier-code) [reply src/solver/utils/filename.h]: Done

File: src/solver/utils/filename.h (hunk 15 above)
Comment (guilpier-code) [reply src/solver/utils/filename.h]: you mean private inheritance ?

File: src/solver/utils/filename.h (hunk 16 above)
Comment (guilpier-code) [reply src/solver/utils/filename.h]: Done

File: src/solver/utils/filename.h (hunk 15 above)
Comment (flomnes) [reply src/solver/utils/filename.h]: These classes shouldn't be exposed in the public API. They should only be exposed inside a private header.

File: src/solver/utils/filename.h (hunk 17 above)
Comment (guilpier-code) [reply src/solver/utils/filename.h]: Done : nothing changed for these 2 functions

**Comment by flomnes** [comment]
TODO
- [ ] Update file **docs/reference-guide/13-file-format.md** to indicate this change in the output format in upcoming version 8.4.2
- [x] Resolve merge conflicts with the base branch

File: src/solver/optimisation/opt_optimisation_lineaire.cpp (hunk 13 above)
Comment (guilpier-code) [reply src/solver/optimisation/opt_optimisation_lineaire.cpp]: Not sure **range** it is better than **period**

File: src/solver/utils/filename.h (hunk 15 above)
Comment (guilpier-code) [reply src/solver/utils/filename.h]: Done

File: src/solver/optimisation/opt_optimisation_lineaire.cpp (hunk 13 above)
Comment (guilpier-code) [reply src/solver/optimisation/opt_optimisation_lineaire.cpp]: Done : **opt_period_as_string** ==> **optPeriodStringGenerator**

**Pull request merged**